From 2e9ac120fea4b0cd62c4541113003ae803f88193 Mon Sep 17 00:00:00 2001
From: Gauvain 'GovanifY' Roussel-Tarbouriech <gauvain@govanify.com>
Date: Sat, 19 Mar 2022 02:24:37 +0100
Subject: [PATCH] frontend: add command line options

This adds the following options:

-f, --fullscreen: loads the emulator in fullscreen
-h, --help: list the commands
-v, --version: prints melonDS's version

past behaviour with nds+gba roms preloading is kept
---
 src/frontend/qt_sdl/main.cpp | 28 +++++++++++++++++++---------
 1 file changed, 19 insertions(+), 9 deletions(-)

diff --git a/src/frontend/qt_sdl/main.cpp b/src/frontend/qt_sdl/main.cpp
index 00135b40b..d3f3c0b24 100644
--- a/src/frontend/qt_sdl/main.cpp
+++ b/src/frontend/qt_sdl/main.cpp
@@ -35,6 +35,7 @@
 #include <QKeyEvent>
 #include <QMimeData>
 #include <QVector>
+#include <QCommandLineParser>
 #ifndef _WIN32
 #include <QSocketNotifier>
 #include <unistd.h>
@@ -2993,6 +2994,22 @@ bool MelonApplication::event(QEvent *event)
 
 int main(int argc, char** argv)
 {
+    MelonApplication melon(argc, argv);
+
+    QCoreApplication::setApplicationName("melonDS");
+    QCoreApplication::setApplicationVersion(MELONDS_VERSION);
+
+    QCommandLineParser cmdargs;
+    cmdargs.setApplicationDescription("Nintendo DS and DSi emulator");
+    cmdargs.addHelpOption();
+    cmdargs.addVersionOption();
+    QCommandLineOption fullscreen({{"f","fullscreen"}, "Starts the emulator in fullscreen."});
+    cmdargs.addOption(fullscreen);
+    cmdargs.addPositionalArgument("roms", "ROMs to load, optionally.", "[roms...]");
+
+    cmdargs.process(melon);
+    QStringList roms = cmdargs.positionalArguments();
+
     srand(time(nullptr));
 
     printf("melonDS " MELONDS_VERSION "\n");
@@ -3000,8 +3017,6 @@ int main(int argc, char** argv)
 
     Platform::Init(argc, argv);
 
-    MelonApplication melon(argc, argv);
-
     // http://stackoverflow.com/questions/14543333/joystick-wont-work-using-sdl
     SDL_SetHint(SDL_HINT_JOYSTICK_ALLOW_BACKGROUND_EVENTS, "1");
 
@@ -3112,14 +3127,9 @@ int main(int argc, char** argv)
 
     QObject::connect(&melon, &QApplication::applicationStateChanged, mainWindow, &MainWindow::onAppStateChanged);
 
-    if (argc > 1)
-    {
-        QString file = argv[1];
-        QString gbafile = "";
-        if (argc > 2) gbafile = argv[2];
+    if (roms.size()) mainWindow->preloadROMs(QString(roms.at(0)), (roms.size() > 1) ? QString(roms.at(1)) : "");
 
-        mainWindow->preloadROMs(file, gbafile);
-    }
+    if(cmdargs.isSet(fullscreen)) emuThread->windowFullscreenToggle();
 
     int ret = melon.exec();
 
